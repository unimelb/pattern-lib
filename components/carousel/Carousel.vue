<template>
  <section
    class="carousel"
    aria-label="Stories and events Carousel">
    <div
      v-if="stories.length"
      class="carousel__slider">
      <div class="carousel__prevent-click" />
      <slider
        ref="slider"
        :options="options"
        @slide="slide">
        <slideritem
          v-for="(slideItem, index) in stories"
          :key="index">
          <div
            :style="image[index]"
            :aria-label="slideItem.altText"
            class="carousel__image" />
        </slideritem>
      </slider>
    </div>
    <div
      class="carousel__panel">
      <div class="carousel__progress-bar">
        <div
          :style="{ width: progressBarWidth + '%' }"
          class="carousel__progress-bar-status" />
      </div>
      <div class="carousel__panel-container">
        <div
          class="carousel__aside"
          role="navigation"
          aria-label="Stories">
          <ul
            class="carousel__navigation"
            role="menu">
            <li
              v-for="(story, index) in storiesData"
              :key="index"
              class="carousel__navigation-item list-reset"
              role="menuitem">
              <a
                :class="{ 'carousel__navigation-link--active': story.isActive }"
                class="carousel__navigation-link"
                href="javascript: void(0);"

                @click="moveToStory(index)"
                @focusin="stopSliding"
                @keydown.enter="moveToStory(index)"
                @keydown.space="moveToStory(index)">
                {{ story.title | truncate(48) }}
              </a>
            </li>
          </ul>
          <div class="carousel__controls">
            <div
              :aria-label="'Previous: ' + prevTitle"
              class="carousel__controls-item"
              tabindex="0"
              role="button"
              @click="prevStory"
              @keydown.enter="prevStory"
              @keydown.space="prevStory"
              @focusin="stopSliding">
              <SvgIcon
                name="chevron-left"
                width="16"
                height="16"
                aria-hidden="true" />
            </div>
            <div
              v-if="!paused"
              tabindex="0"
              class="carousel__controls-item carousel__controls-stop-pause"
              data-message="Play"
              role="button"
              aria-label="Stop carousel"
              @click="stopSliding"
              @keydown.enter="stopSliding"
              @keydown.space="stopSliding">
              <SvgIcon
                name="pause"
                width="16"
                height="16"
                aria-hidden="true" />
            </div>
            <div
              v-if="paused"
              tabindex="0"
              class="carousel__controls-item"
              data-message="Pause"
              role="button"
              aria-label="Play carousel"
              @click="startSliding"
              @focusin="stopSliding"
              @keydown.enter="startSliding"
              @keydown.space="startSliding">
              <SvgIcon
                name="play"
                width="16"
                height="16"
                aria-hidden="true" />
            </div>
            <div
              :aria-label="'Next: ' + nextTitle"
              class="carousel__controls-item"
              tabindex="0"
              role="button"
              @click="nextStory"
              @keydown.enter="nextStory"
              @keydown.space="nextStory"
              @focusin="stopSliding">
              <SvgIcon
                name="chevron-right"
                width="16"
                height="16"
                aria-hidden="true" />
            </div>
          </div>
        </div>
        <div
          :aria-live="!paused ? 'off' : 'polite'"
          class="carousel__content"
          role="navigation"
          aria-label="Stories">
          <div class="carousel__story">
            <h2 class="carousel__title">
              <a
                :href="selectedItem.buttonHref"
                class="carousel__title-link"
                @focusin="stopSliding">
                {{ selectedItem.title | truncate(48) }}
              </a>
            </h2>
            <p class="carousel__description">
              {{ selectedItem.description | truncate(240, '...') }}
            </p>
            <div
              @focusin="stopSliding">
              <ButtonIcon
                :href="selectedItem.buttonHref"
                inverted
                width="wide"
                role="button"
                size="sml">
                {{ selectedItem.buttonText }}
              </ButtonIcon>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</template>

<script>
import { slider, slideritem } from 'vue-concise-slider';
import SvgIcon from '../icons/SvgIcon.vue';
import ButtonIcon from '../buttons/ButtonIcon.vue';

import { TIMER_1100, TIMER_200 } from '../../constants/timers';

export default {
  components: {
    slider,
    slideritem,
    SvgIcon,
    ButtonIcon,
  },
  filters: {
    truncate(value, limit, ellipsis = '') {
      const startFromIndex = 0;

      if (!value) {
        return '';
      }

      if (value.length < limit) {
        return value;
      }

      return `${value.substring(startFromIndex, limit)}${ellipsis}`;
    },
  },
  props: {
    stories: {
      type: Array,
      default: () => [],
      validator: (stories) => {
        const maxStories = 3;

        return stories.length <= maxStories;
      },
      required: true,
    },
    timing: {
      type: String,
      default: 'short',
    },
  },
  data() {
    const timing = {
      medium: 16000,
      long: 20000,
      def: 12000,
    };
    const autoplay = this.timing && timing[this.timing] ? timing[this.timing] : timing.def;

    return {
      prevTitle: this.stories && this.stories[0] ? this.stories[this.stories.length - 1].title : {},
      nextTitle: this.stories && this.stories[1] ? this.stories[1].title : {},
      selectedItem: this.stories && this.stories[0] ? this.stories[0] : {},
      selectedIndex: 0,
      openState: false,
      progressBarWidth: 0,
      countTime: 0,
      interval: null,
      paused: false,
      autoplay,
      options: {
        effect: 'fade',
        currentPage: 0,
        pagination: false,
        thresholdDistance: 0, // Sliding distance threshold
        thresholdTime: 0, // Sliding time threshold decision
        grabCursor: false, // Scratch style
        speed: 300, // Sliding speed
        timingFunction: 'ease', // Sliding mode
        loop: true, // Infinite loop
        autoplay, // Auto play[ms]
        freeze: true,
      },
      image: {},
    };
  },
  computed: {
    storiesData() {
      const data = [];

      this.stories.forEach((item, index) => {
        data.push({
          ...item,
          isActive: index === this.selectedIndex,
        });
      });

      return data;
    },
  },
  created() {
    const startFromPositionIndex = 0;

    this.displayImageAtPosition(startFromPositionIndex);
  },
  mounted() {
    this.loadImg();
    setTimeout(() => {
      this.actionProgressBar();
    }, TIMER_1100);
  },
  methods: {
    displayImageAtPosition(index) {
      this.image[index] = {
        backgroundImage: `url(${this.stories[index].src})`,
        backgroundPosition: this.stories[index].imagePosition ? this.stories[index].imagePosition : 'center',
      };
    },
    loadImg() {
      setTimeout(() => {
        for (let i = 1; i < this.stories.length; i += 1) {
          this.displayImageAtPosition(i);
        }
      }, TIMER_1100);
    },
    slide(slide) {
      if (this.stories[slide.currentPage]) {
        this.selectedItem = this.stories[slide.currentPage];
        this.selectedIndex = slide.currentPage;
        this.prevTitle = this.stories[this.selectedIndex - 1] ? this.stories[this.selectedIndex - 1].title : this.stories[this.stories.length - 1].title;
        this.nextTitle = this.stories[this.selectedIndex + 1] ? this.stories[this.selectedIndex + 1].title : this.stories[0].title;
      }
      this.clearFrame();
    },
    stopSliding() {
      this.$refs.slider.$emit('autoplayStop');
      clearInterval(this.interval);
      this.paused = true;
    },
    startSliding() {
      this.$refs.slider.$emit('autoplayStart', this.autoplay - this.countTime);
      this.paused = false;
      this.actionProgressBar();
    },
    frame() {
      const initialWidth = 5;
      const percent = 100;

      this.progressBarWidth = Math.round((this.countTime / this.autoplay) * percent) + initialWidth;
      this.countTime += 200;
    },
    actionProgressBar() {
      clearInterval(this.interval);
      this.interval = null;
      this.interval = setInterval(() => this.frame(), TIMER_200);
    },
    clearFrame() {
      if (!this.paused) {
        this.$refs.slider.$emit('autoplayStop');
        this.$refs.slider.$emit('autoplayStart', this.autoplay);
      }
      this.progressBarWidth = 0;
      this.countTime = 0;
    },
    prevStory() {
      this.$refs.slider.$emit('slidePre');
      this.progressBarWidth = 0;
      this.countTime = 0;
    },
    nextStory() {
      this.$refs.slider.$emit('slideNext');
      this.progressBarWidth = 0;
      this.countTime = 0;
    },
    moveToStory(storyIndex) {
      this.displayImageAtPosition(storyIndex);
      this.selectedItem = this.stories[storyIndex];
      this.selectedIndex = storyIndex;
      this.$refs.slider.$emit('slideTo', storyIndex);
      document.querySelector('.carousel__title').focus();
    },
  },
};
</script>
