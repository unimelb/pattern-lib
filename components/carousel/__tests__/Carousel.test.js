import {
  shallow,
  mount,
} from 'vue-test-utils';
import { axe, toHaveNoViolations } from 'jest-axe';
import sinon from 'sinon';
import Carousel from '../Carousel.vue';

expect.extend(toHaveNoViolations);

const storiesMock = [{
  src: 'test1',
  imagePosition: 'center',
  altText: 'test',
  title: 'Story 1',
  description: 'Story 2 description',
  buttonText: 'Register',
  buttonHref: 'google.com',
},
{
  src: 'test2',
  imagePosition: 'center',
  altText: 'test',
  title: 'Story 2',
  description: 'Story 2 description',
  buttonText: 'Login',
  buttonHref: 'google.com',
},
];

describe('Carousel', () => {
  it('should match snapshot', () => {
    const result = shallow(Carousel, {
      propsData: {
        stories: storiesMock,
      },
    }).element;

    expect(result).toMatchSnapshot();
  });

  it('should have default props and correct types stories', () => {
    const wrapper = shallow(Carousel, {
      propsData: {
        stories: storiesMock,
      },
    });
    const {
      stories,
    } = wrapper.vm.$options.props;

    expect(stories.type).toBe(Array);
    expect(wrapper.props().stories).toEqual(storiesMock);
  });

  it('should accept stories prop with correct type', () => {
    const wrapper = shallow(Carousel, {
      propsData: {
        stories: storiesMock,
      },
    });

    expect(typeof wrapper.props().stories).toBe('object');
    expect(wrapper.props().stories).toBe(storiesMock);
  });

  it('should render story in slider from prop', () => {
    const wrapper = mount(Carousel, {
      propsData: {
        stories: storiesMock,
      },
    });
    expect(wrapper.props().stories.length).toBe(2);

    expect(wrapper.find('.carousel__image').attributes().style).toBe('background-image: url(test1); background-position: center;');
  });

  it('should set active to first story as default', () => {
    const wrapper = mount(Carousel, {
      propsData: {
        stories: storiesMock,
      },
    });
    expect(wrapper.vm.selectedIndex).toBe(0);
    expect(wrapper.find('.carousel__image').exists()).toBe(true);
  });

  it('should render default story title and description', () => {
    const wrapper = mount(Carousel, {
      propsData: {
        stories: storiesMock,
      },
    });
    expect(wrapper.find('.carousel__title').text()).toBe('Story 1');
    expect(wrapper.find('.carousel__description').text()).toBe('Story 2 description');
  });

  it('should render stories titles menu', () => {
    const wrapper = mount(Carousel, {
      propsData: {
        stories: storiesMock,
      },
    });
    expect(wrapper.findAll('.carousel__navigation-link').at(0).text()).toBe('Story 1');
    expect(wrapper.findAll('.carousel__navigation-link').at(1).text()).toBe('Story 2');
  });

  it('should have click/enter event on stories menu title', () => {
    const moveToStory = sinon.stub();
    const wrapper = shallow(Carousel, {
      propsData: {
        stories: storiesMock,
      },
    });

    wrapper.setMethods({
      moveToStory,
    });
    wrapper.find('.carousel__navigation-link').trigger('click');
    expect(moveToStory.called).toBe(true);
    wrapper.find('.carousel__navigation-link').trigger('keydown.enter');
    expect(moveToStory.called).toBe(true);
  });

  it('should change story on click event on stories menu title', () => {
    const wrapper = shallow(Carousel, {
      propsData: {
        stories: storiesMock,
      },
      attachToDocument: true,
    });

    wrapper.findAll('.carousel__navigation-link').at(1).trigger('click');
    expect(wrapper.vm.selectedIndex).toBe(1);
  });

  it('should set "active" state to active story', () => {
    const wrapper = shallow(Carousel, {
      propsData: {
        stories: storiesMock,
      },
    });

    expect(wrapper.vm.storiesData[0].isActive).toBe(true);
    expect(wrapper.find('.carousel__navigation-link').classes()).toContain('carousel__navigation-link--active');
  });

  it('should set "active" state to story on menu title click', () => {
    const wrapper = shallow(Carousel, {
      propsData: {
        stories: storiesMock,
      },
      attachToDocument: true,
    });

    wrapper.findAll('.carousel__navigation-link').at(1).trigger('click');
    expect(wrapper.vm.storiesData[1].isActive).toBe(true);
    expect(wrapper.findAll('.carousel__navigation-link').at(1).classes()).toContain('carousel__navigation-link--active');
  });

  it('should have click/enter event on stop/pause control', () => {
    const stopSliding = sinon.stub();
    const startSliding = sinon.stub();
    const wrapper = mount(Carousel, {
      propsData: {
        stories: storiesMock,
      },
    });

    wrapper.setMethods({
      stopSliding,
      startSliding,
    });

    const pauseStop = wrapper.find('.carousel__controls-stop-pause');

    pauseStop.trigger('click');
    expect(stopSliding.called).toBe(true);
    wrapper.vm.paused = false;
    pauseStop.trigger('keydown.enter');
    expect(stopSliding.called).toBe(true);
    wrapper.setData({
      paused: true,
    });
    pauseStop.trigger('click');
    expect(startSliding.called).toBe(true);
    pauseStop.trigger('keydown.enter');
    expect(startSliding.called).toBe(true);
  });

  it('should have click/enter event on left/right control', () => {
    const prevStory = sinon.stub();
    const nextStory = sinon.stub();
    const wrapper = mount(Carousel, {
      propsData: {
        stories: storiesMock,
      },
    });

    wrapper.setMethods({
      prevStory,
      nextStory,
    });

    const prev = wrapper.findAll('.carousel__controls-item').at(0);
    const next = wrapper.findAll('.carousel__controls-item').at(2);

    prev.trigger('click');
    expect(prevStory.called).toBe(true);
    prev.trigger('keydown.enter');
    expect(prevStory.called).toBe(true);

    next.trigger('click');
    expect(nextStory.called).toBe(true);
    next.trigger('keydown.enter');
    expect(nextStory.called).toBe(true);
  });

  it('should change slide on left/right controls click', () => {
    const slide = sinon.stub();

    const wrapper = mount(Carousel, {
      propsData: {
        stories: storiesMock,
      },
    });

    wrapper.setMethods({
      slide,
    });

    const prev = wrapper.findAll('.carousel__controls-item').at(0);
    const next = wrapper.findAll('.carousel__controls-item').at(2);

    prev.trigger('click');
    expect(slide.called).toBe(true);
    next.trigger('click');
    expect(slide.called).toBe(true);
  });

  it('Component throws no accessibility violations', (done) => {
    const html = shallow(Carousel, {
      propsData: {
        stories: storiesMock,
      },
    }).html();
    return axe(html).then((response) => {
      expect(response).toHaveNoViolations();
      done();
    });
  });
});
