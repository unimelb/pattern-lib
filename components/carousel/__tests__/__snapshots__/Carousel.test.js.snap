// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel should match snapshot 1`] = `
<section
  aria-label="Stories and events Carousel"
  class="carousel"
>
  <div
    class="carousel__slider"
  >
    <div
      class="carousel__prevent-click"
    />
     
    <!---->
  </div>
   
  <div
    class="carousel__panel"
  >
    <div
      class="carousel__progress-bar"
    >
      <div
        class="carousel__progress-bar-status"
        style="width: 0%;"
      />
    </div>
     
    <div
      class="carousel__panel-container"
    >
      <div
        aria-label="Stories"
        class="carousel__aside"
        role="navigation"
      >
        <ul
          class="carousel__navigation"
          role="menu"
        >
          <li
            class="carousel__navigation-item list-reset"
            role="menuitem"
          >
            <a
              class="carousel__navigation-link carousel__navigation-link--active"
              href="javascript: void(0);"
            >
              
              Story 1
            
            </a>
          </li>
          <li
            class="carousel__navigation-item list-reset"
            role="menuitem"
          >
            <a
              class="carousel__navigation-link"
              href="javascript: void(0);"
            >
              
              Story 2
            
            </a>
          </li>
        </ul>
         
        <div
          class="carousel__controls"
        >
          <div
            aria-label="Previous: Story 2"
            class="carousel__controls-item"
            role="button"
            tabindex="0"
          >
            <!---->
          </div>
           
          <div
            aria-label="Stop carousel"
            class="carousel__controls-item carousel__controls-stop-pause"
            data-message="Play"
            role="button"
            tabindex="0"
          >
            <!---->
          </div>
           
          <!---->
           
          <div
            aria-label="Next: Story 2"
            class="carousel__controls-item"
            role="button"
            tabindex="0"
          >
            <!---->
          </div>
        </div>
      </div>
       
      <div
        aria-label="Stories"
        aria-live="off"
        class="carousel__content"
        role="navigation"
      >
        <div
          class="carousel__story"
        >
          <h2
            class="carousel__title"
          >
            <a
              class="carousel__title-link"
              href="google.com"
            >
              
              Story 1
            
            </a>
          </h2>
           
          <p
            class="carousel__description"
          >
            
            Story 2 description
          
          </p>
           
          <div>
            <!---->
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
